File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/QueryEndAtConstraint.java
package io.capawesome.capacitorjs.plugins.firebase.firestore.classes;

import com.getcapacitor.JSObject;

import io.capawesome.capacitorjs.plugins.firebase.firestore.interfaces.QueryNonFilterConstraint;

public class QueryEndAtConstraint implements QueryNonFilterConstraint {
    private String reference;

    public QueryEndAtConstraint(JSObject queryConstraint) {
        this.reference = queryConstraint.getString("reference");
    }

    public String getReference() {
        return reference;
    }
}

File: packages/firestore/ios/Plugin/Classes/Constraints/QueryEndAtConstraint.swift
import Foundation
import FirebaseFirestore
import Capacitor

@objc public class QueryEndAtConstraint: NSObject, QueryNonFilterConstraint {
    private var type: String
    private var reference: String

    public init(_ queryConstraint: JSObject) {
        self.type = queryConstraint["type"] as! String
        self.reference = queryConstraint["reference"] as! String
    }

    public func toQuery(_ query: Query, completion: @escaping (Query, Error?) -> Void) {
        Firestore.firestore().document(reference).getDocument() { documentSnapshot, error in
            if let error = error {
                completion(query, error)
            } else {
                var newQuery: Query
                switch self.type {
                case "endAt":
                    newQuery = query.end(atDocument: documentSnapshot!)
                default:
                    newQuery = query.end(beforeDocument: documentSnapshot!)
                }
                completion(newQuery, nil)
            }
        }
    }
}
